scripts/add_electricity.py

```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
                                       "'{tech}' is not implemented, yet. "
                                       "Only OCGT, CCGT and nuclear are allowed at the moment.")
 
+def map_ppls_to_bus(xy, buses, tech):
+    # map ppl with coordinates xy \in R^2 to closest bus (in buses) of the network;
+    idx = np.sqrt((n.buses.loc[buses].x-xy[1])**2+(n.buses.loc[buses].y-xy[0])**2).argmin()
+    return n.buses.loc[buses].iloc[idx].name + ' ' + tech
+
+def attach_OPSD_ppls_for_country(country, tech_map=None):
+    # retrieve databunde from OPSD (Open Power System Data) and map ppls to closest potential bus
 
+    if tech_map is None:
+        tech_map = (snakemake.config['electricity']
+                    .get('include_renewable_capacities_from_OPSD', {}))
+
+    capacities = (ppm.data.OPSD_VRE_country(country).powerplant.convert_country_to_alpha2()
+                  .astype({'Capacity': 'float'})
+                  .groupby(['lat','lon','Fueltype','Technology']) # reduce input data
+                  .agg({'Capacity': 'sum'})
+                  .query('Capacity > 0.1') # only capacities above 100kW
+                  .reset_index())
+
+    # distinguish between onshore and offshore, if given:
+    capacities.Fueltype = (capacities[['Fueltype','Technology']]
+                           .apply(lambda b: b.Technology if b.Technology in ['Onshore', 'Offshore']
+                                  else b.Fueltype, axis=1))
+
+    # retrieve a mapping ppl->bus if locations (lat/lon) for ppls are given:
+    if ~capacities[['lat','lon']].isna().stack().all():
+        buses = n.buses[n.buses.country==country].index
+        
+        mapping = pd.Series(dtype='float')
+        for ppm_fueltype, techs in tech_map.items():
+            for tech in techs:
+                tech_i = n.generators.query('carrier in @tech and bus in @buses').index
+
+                # fix available buses in map_ppls_to_bus function:
+                map_ppls = partial(map_ppls_to_bus, buses=n.generators.loc[tech_i].bus, tech=tech)
+
+                # parallel assignment of ppls to buses; each process is independent to speed up
+                pool_buses = multiprocessing.Pool(processes=4)
+                mapping = (mapping.append(
+                    pd.Series((capacities.query('Fueltype in @ppm_fueltype').Capacity/len(techs)).values,
```

Comment:
Here you divide by `len(techs)`. This assumes that all techs are at all buses, right?. However, there are cases where you have only onwind or only offwind (not both) at a bus.